### PR TITLE
Added code to handle delete session procedure 

### PR DESCRIPTION
**Pull request by rsarwad**
"[lte][agw]: 
Added code to handle delete session procedure and added routing logic at mme to s11 message to either sgw_s8 task or to spgw_task, based on federated mode of configuration
## Summary
1. Added code at sgw_s8 task to handle Delete Session Request message
2. Built the gRPC message
3. Added code at sgw_s8 task to handle Delete Session response and delete the contexts and OVS tunnels
4. Added routing logic at mme based federated mode of configuration
5. Moved the routing logic from S6Client.cpp to common file, so that it can be accessed by both mme_app and s6a tasks.

## Test Plan
1. Executed s1ap sanity test suite
2. Executed basic attach and detach scenario over S8 task in Tera VM setup

## Additional Information
This PR include changes of handling of both modify bearer request and delete session procedure.
To view the handling of delete session procedure and routing logic at mme, is available in PR, 
https://github.com/rsarwad/magma/pull/9

This PR is dependent on PR, https://github.com/magma/magma/pull/6046 to insert ovs rules.

### CODE REVIEW

File: lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c [hunk 1]
```diff
@@ -84,6 +84,10 @@
 extern task_zmq_ctx_t mme_app_task_zmq_ctx;
 extern int pdn_connectivity_delete(emm_context_t* emm_context, pdn_cid_t pid);
 
+static void send_s11_modify_bearer_request(
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:87]: if this is intended for S8 task only then I suggest to use something like `send_modify_bearer_request_to_S8task`.

Comment (ulaskozat) [reply lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:87]: never mind I have missed that this is splitting the request among S8 and and SPGW tasks.

Comment (rsarwad) [reply lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:87]: No problem

File: lte/gateway/c/oai/tasks/mme_app/mme_app_itti_messaging.c [hunk 2]
```diff
@@ -373,20 +375,23 @@ int mme_app_send_s11_create_session_req(
...
-      "Sending S11 CREATE SESSION REQ message to SPGW for "
-      "ue_id " MME_UE_S1AP_ID_FMT "\n",
-      ue_mm_context->mme_ue_s1ap_id);
-  if ((send_msg_to_task(&mme_app_task_zmq_ctx, TASK_SPGW, message_p)) !=
-      RETURNok) {
-    OAILOG_ERROR_UE(
+  int mode =
+      match_fed_mode_map((char*) session_request_p->imsi.digit, LOG_MME_APP);
+  if (mode == S8_SUBSCRIBER) {
+    OAILOG_INFO_UE(
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_itti_messaging.c:381]: shall we make this debug level?

File: lte/gateway/c/oai/tasks/mme_app/mme_app_itti_messaging.c [hunk 3]
```diff
@@ -373,20 +375,23 @@ int mme_app_send_s11_create_session_req(
...
         LOG_MME_APP, ue_mm_context->emm_context._imsi64,
-        "Failed to send S11 CREATE SESSION REQ message to SPGW for "
+        "Sending s11 create session req message to SGW_s8 task for "
         "ue_id " MME_UE_S1AP_ID_FMT "\n",
         ue_mm_context->mme_ue_s1ap_id);
-    OAILOG_FUNC_RETURN(LOG_MME_APP, RETURNerror);
+    send_msg_to_task(&mme_app_task_zmq_ctx, TASK_SGW_S8, message_p);
+    ue_mm_context->pdn_contexts[pdn_cid]->route_s11_messages_to_s8_task = true;
+  } else {
+    OAILOG_INFO_UE(
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_itti_messaging.c:389]: debug level?

File: lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c [hunk 4]
```diff
@@ -4086,3 +4083,30 @@ void mme_app_handle_modify_bearer_rsp(
...
+    ue_mm_context_t* ue_context_p, pdn_context_t* pdn_context_p,
+    MessageDef* message_p) {
+  OAILOG_FUNC_IN(LOG_MME_APP);
+  Imsi_t imsi = {0};
+  IMSI64_TO_STRING(
+      ue_context_p->emm_context._imsi64, (char*) (&imsi.digit),
+      ue_context_p->emm_context._imsi.length);
+
+  if (pdn_context_p->route_s11_messages_to_s8_task) {
+    OAILOG_INFO_UE(
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:4097]: debug level?

File: lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c [hunk 5]
```diff
@@ -4086,3 +4083,30 @@ void mme_app_handle_modify_bearer_rsp(
...
+        LOG_MME_APP, ue_context_p->emm_context._imsi64,
+        "Sending S11 modify bearer req message to SGW_s8 task for "
+        "ue_id " MME_UE_S1AP_ID_FMT "\n",
+        ue_context_p->mme_ue_s1ap_id);
+    send_msg_to_task(&mme_app_task_zmq_ctx, TASK_SGW_S8, message_p);
+  } else {
+    OAILOG_INFO_UE(
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:4104]: debug level?

File: lte/gateway/c/oai/lib/s8_proxy/s8_client_api.cpp [hunk 6]
```diff
@@ -101,6 +103,56 @@ static void get_paa_from_proto_msg(
...
+    }
+  } else {
+    OAILOG_ERROR_UE(
+        LOG_SGW_S8, imsi64,
+        "Received gRPC error for delete session response for "
+        "context_teid " TEID_FMT "\n",
+        context_teid);
+    s8_delete_session_rsp->cause = REMOTE_PEER_NOT_RESPONDING;
+  }
+  OAILOG_INFO_UE(
```
Comment (ulaskozat) [inline lte/gateway/c/oai/lib/s8_proxy/s8_client_api.cpp:140]: DEBUG level?

File: lte/gateway/c/oai/lib/s8_proxy/s8_client_api.cpp [hunk 7]
```diff
@@ -371,6 +440,36 @@ static void convert_proto_msg_to_itti_csr(
...
+    s5_response->cause = REQUEST_ACCEPTED;
+  }
+  OAILOG_FUNC_OUT(LOG_SGW_S8);
+}
+
+void send_s8_delete_session_request(
+    imsi64_t imsi64, Imsi_t imsi, teid_t sgw_s11_teid, teid_t pgw_s5_teid,
+    ebi_t bearer_id) {
+  OAILOG_FUNC_IN(LOG_SGW_S8);
+  OAILOG_INFO_UE(
```
Comment (ulaskozat) [inline lte/gateway/c/oai/lib/s8_proxy/s8_client_api.cpp:465]: DEBUG level?

File: lte/gateway/c/oai/tasks/mme_app/mme_app_detach.c [hunk 8]
```diff
@@ -115,10 +116,25 @@ void mme_app_send_delete_session_request(
...
-  OAILOG_INFO(
-      LOG_MME_APP, "Send Delete session Req for teid " TEID_FMT "\n",
-      ue_context_p->mme_teid_s11);
+  Imsi_t imsi = {0};
+  IMSI64_TO_STRING(
+      ue_context_p->emm_context._imsi64, (char*) (&imsi.digit),
+      ue_context_p->emm_context._imsi.length);
+
+  if (ue_context_p->pdn_contexts[cid]->route_s11_messages_to_s8_task) {
+    OAILOG_INFO_UE(
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_detach.c:125]: debug level?

File: lte/gateway/c/oai/tasks/mme_app/mme_app_detach.c [hunk 9]
```diff
@@ -115,10 +116,25 @@ void mme_app_send_delete_session_request(
...
+        LOG_MME_APP, ue_context_p->emm_context._imsi64,
+        "Send delete session Req for teid to sgw_s8 task " TEID_FMT "\n",
+        ue_context_p->mme_teid_s11);
+    send_msg_to_task(&mme_app_task_zmq_ctx, TASK_SGW_S8, message_p);
+  } else {
+    OAILOG_INFO_UE(
```
Comment (ulaskozat) [inline lte/gateway/c/oai/tasks/mme_app/mme_app_detach.c:131]: debug level?

File: lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c (hunk 4 above)
Comment (rsarwad) [reply lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:4097]: My understanding on info level would capture the call flow; So I had logged with info, to know where s11 messages would be routed through.
Let me know your thoughts on same

File: lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c (hunk 5 above)
Comment (rsarwad) [reply lte/gateway/c/oai/tasks/mme_app/mme_app_bearer.c:4104]: same as above

**Review by ulaskozat — APPROVED** [review]
LGTM.

**Comment by ulaskozat** [comment]
This PR is passing integ tests in dev environment and CI failure looks like a server connection issue. Going to merge it.

**Pull request merged**